File: algorithms/insertionsort/InsertionSort.java
import java.util.Arrays;

class InsertionSort {

	// ascending
	private static void insertion_sort_asc(int [] A) {
		int i,j,t;
		for(i = 0; i < A.length; i++) {
			j = i;
			while(j > 0 && A[j-1] > A[j]) {
				t = A[j];
				A[j] = A[j-1];
				A[j-1] = t;
				j--;
			}
		}
	}

	// descending
	private static void insertion_sort_desc(int [] A) {
		int i,j,t;
		for(i = 0; i < A.length; i++) {
			j = i;
			while(j > 0 && A[j-1] < A[j]) {
				t = A[j];
				A[j] = A[j-1];
				A[j-1] = t;
				j--;
			}
		}
	}

	public static void main(String[] args) {
		int [] A = {8, 3, 1, 6, 10, 2, 9, 5, 7, 4};
		insertion_sort_asc(A);
		System.out.println("Ascending: " + Arrays.toString(A));
		insertion_sort_desc(A);
		System.out.println("Descending: " + Arrays.toString(A));
	}

}

File: algorithms/insertionsort/insertionsort.lua
-- ascending
function insertion_sort_asc(A)
	local i, j
	for i = 2, #A do
		j = i
		while j > 1 and A[j-1] > A[j] do
			A[j-1], A[j] = A[j], A[j-1]
			j = j - 1
		end
	end
end

-- descending
function insertion_sort_desc(A)
	local i, j
	for i = 2, #A do
		j = i
		while j > 1 and A[j-1] < A[j] do
			A[j-1], A[j] = A[j], A[j-1]
			j = j - 1
		end
	end
end


-- main
A = { 8, 3, 1, 6, 10, 2, 9, 5, 7, 4 }
print("Original:   [" .. table.concat(A, ", ") .. "]")
insertion_sort_asc(A)
print("Ascending:  [" .. table.concat(A, ", ") .. "]")
insertion_sort_desc(A)
print("Descending: [" .. table.concat(A, ", ") .. "]")